proxy/src/bind.rs

```diff
@@ -196,7 +207,61 @@ where
     type BindError = ();
 
     fn bind(&self, addr: &SocketAddr) -> Result<Self::Service, Self::BindError> {
-        Ok(self.bind.bind_service(addr, self.protocol))
+        Ok(self.bind.bind_service(addr, &self.protocol))
     }
 }
 
+// ===== impl Protocol =====
+
+
+impl<'a, B> From<&'a http::Request<B>> for Protocol {
+    fn from(req: &'a http::Request<B>) -> Protocol {
+        if req.version() == http::Version::HTTP_2 {
+            return Protocol::Http2
+        }
+
+        if req.extensions().get::<h1::AuthorityRewriting>() ==
+            Some(&h1::AuthorityRewriting::SoOriginalDst)
+        {
+            return Protocol::Http1(Host::NoAuthority);
+        }
+
+        // If the request has an authority part, use that as the host part of
+        // the key for an HTTP/1.x request.
+        let host = req.uri().authority_part()
+                .cloned()
+                .or_else(|| {
+                    // No authority part in the request URI, so try and use the
+                    // Host: header value, if one is present and it can be parsed
+                    // as an Authority.
+                        req.headers().get(header::HOST)
```

Comment:
Checking the `Host` header probably isn't needed here, since at the accept point, the h1 server injects a `Host` header into the `Uri` if needed.